### PR TITLE
Adding --preinstall flag to brew update

### PR DESCRIPTION
**Pull request by Anipik**
Fixes https://dev.azure.com/dnceng/internal/_build/results?buildId=851404&view=logs&j=8fda0d10-4a57-56af-7add-7281436b36e1&t=d354d20b-3af3-59c3-d621-c743753ae5da

brew update fetchs the newest version of Homebrew and all formulae from GitHub using git and perform any necessary migrations. There was problem fetching those issues.
preinstall helps 
```
--preinstall: Run on auto-updates (e.g. before brew install). Skips some slower steps.
```

i looked at the what we were differently in dotnet runtime to find the --preinstall flag.

### CODE REVIEW

**Comment by Anipik** [comment]
https://dev.azure.com/dnceng/internal/_build/results?buildId=853357&view=results 

Relative internal build

File: eng/install-native-dependencies.sh [hunk 1]
```diff
@@ -10,15 +10,9 @@ if [ "$1" = "Linux" ]; then
...
 elif [ "$1" = "OSX" ]; then
-    brew update
-    if [ "$?" != "0" ]; then
-        exit 1;
-    fi
-    brew install icu4c openssl
-    if [ "$?" != "0" ]; then
-        exit 1;
-    fi
-    brew link --force icu4c
```
Comment (ericstj) [inline eng/install-native-dependencies.sh]: Do we no longer need this `--force icu4c`?

File: eng/BrewFile [hunk 2]
```diff
@@ -0,0 +1,8 @@
+brew "autoconf"
+brew "automake"
+brew "cmake"
+brew "icu4c"
+brew "libtool"
+brew "openssl@1.1"
```
Comment (ericstj) [inline eng/BrewFile]: I see we specified 1.1 here as well as added other packages.  What's the minimal fix here, since this is servicing.

Comment (Anipik) [reply eng/BrewFile]: removed the extras to minimal change.

**Comment by ViktorHofer** [comment]
Which commits did you cherry-pick from dotnet/runtime? cc @akoeplinger

File: eng/install-native-dependencies.sh [hunk 3]
```diff
@@ -10,14 +10,9 @@ if [ "$1" = "Linux" ]; then
...
-    if [ "$?" != "0" ]; then
-        exit 1;
-    fi
-    brew install icu4c openssl
-    if [ "$?" != "0" ]; then
-        exit 1;
-    fi
+    engdir=$(dirname "${BASH_SOURCE[0]}")
+    brew update --preinstall
+    brew bundle --no-upgrade --no-lock --file "${engdir}/BrewFile"
```
Comment (akoeplinger) [inline eng/install-native-dependencies.sh]: nit: the file is called `Brewfile` in dotnet/runtime, not `BrewFile`. would be nice to match the casing.

File: eng/install-native-dependencies.sh [hunk 4]
```diff
@@ -10,14 +10,9 @@ if [ "$1" = "Linux" ]; then
...
 elif [ "$1" = "OSX" ]; then
-    brew update
-    if [ "$?" != "0" ]; then
-        exit 1;
-    fi
-    brew install icu4c openssl
-    if [ "$?" != "0" ]; then
-        exit 1;
-    fi
+    engdir=$(dirname "${BASH_SOURCE[0]}")
```
Comment (akoeplinger) [inline eng/install-native-dependencies.sh]: the shebang at the top of this file should be `#!/usr/bin/env bash` to make sure `${BASH_SOURCE[0]}` works (it is that way in dotnet/runtime master).

**Review by akoeplinger — DISMISSED** [review]

**Comment by ericstj** [comment]
It'd be nice if the title of this PR and it's body expressed the problem and why this change fixes it.  I'm still a bit lost connecting the dots here.

**Comment by Anipik** [comment]
> It'd be nice if the title of this PR and it's body expressed the problem and why this change fixes it. I'm still a bit lost connecting the dots here.

the error does not give any much info other than brew update showing signs of failure. Its not easily reproducible on local machine. I used runtime copy of the file as a reference to fix this.

**Comment by akoeplinger** [comment]
Hm, there should be no difference except being a bit slower without the `--preinstall` flag. If you want to truly avoid the autoupgrades then using the brew bundle approach like you had previously is preferred.

**Comment by Anipik** [comment]
> Hm, there should be no difference except being a bit slower without the --preinstall flag. If you want to truly avoid the autoupgrades then using the brew bundle approach like you had previously is preferred.

the OSX legs passes here https://dev.azure.com/dnceng/internal/_build/results?buildId=855185&view=results
but fails without --preinstall https://dev.azure.com/dnceng/internal/_build/results?buildId=855186&view=results

**Comment by akoeplinger** [comment]
Ok I guess if you want to go with the minimal change that works then this is fine :) Ideally I think we should still copy what we have in dotnet/runtime master to avoid upgrading dependencies during the build at all.

**Review by akoeplinger — APPROVED** [review]

**Pull request merged**